[jvm-packages] Scala implementation of the Rabit tracker.

xydrolase opened the pull request:
## Motivation

The Java implementation of RabitTracker in `xgboost4j` [depends on the Python script](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/RabitTracker.java#L113) `tracker.py` in [dmlc-core](https://github.com/dmlc/dmlc-core) to handle all socket connections / loggings. 

The reliance on Python code has a few weaknesses:
- It makes `xgboost4j-spark` and `xgboost4j-flink`, which use RabitTracker, more susceptible to random failures on worker nodes due to Python versions. 
- It increases difficulty for debugging tracker-related issues.
- Since the Python code handles all socket connection logic, it is difficult to handle timeout due to connection issues, and thus the tracker may hang indefinitely if the workers fail to connect due to networking issues.

To address the above issues, this PR was created to introduce a pure Scala implementation of the RabitTracker, that is interchangeable with the Java implementation at interface level, but with the Python dependency completely removed. 

The implementation was tested in a Spark cluster running on YARN with up to 16 distributed workers. More thorough tests (local mode, more nodes etc.) of this PR is still WIP.
## Implementation details

The Scala version of `RabitTracker` replicates the functionality of the `RabitTracker` class in `tracker.py`, that is, to handle incoming connections from Rabit clients of the worker nodes, compute the link map and rank for each given worker, and print tracker logging information. 

The tracker handles connections in asynchronous and non-blocking fashion using Akka, and resolves the inter-dependency between worker connections properly.
### Timeouts

The Scala RabitTracker implements timeout logic at multiple entry points.
- `RabitTracker.start()`may time out if the tracker fails to bind to a socket address within certain time limit.
- `RabitTracker.waitFor()` may time out if at least one worker fails to connect to the tracker within certain time limit. This prevents the tracker from hanging forever.
- `RabitTracker.waitFor()` may time out after a given maximum execution time limit.
## Checklist

The following tasks are to be completed:
- [x] Add options to switch between Python-based tracker and Scala-based tracker in `xgboost4j-spark` and `xgboost4j-flink`.
- [x] Refactoring of `RabitTracker.scala` to separate the components into different files.
- [x] Unit tests for individual actors (using `akka-testkit`).
- [x] Test with rabit clients (Allreduce, checkpoint, simulated connection issus.)
- [x] Test in production.

CodingCat commented:
Hi, @xydrolase , thanks for this PR, just one quick question, how to sync scala-based implementation with the python-based one?

xydrolase commented:
@CodingCat I'm not an expert on Rabit so I don't know how frequently the protocols of Rabit changes. Assuming that Rabit changes infrequently, I can manually monitor the changes on the Python implementation, and port them to the Scala code.

CodingCat commented:
I understand that the users are facing difficulties from Rabit and I really appreciate the effort to address the issue

I did not have enough energy to figure out the real reason behind all various symptoms...but I personally prefer a more user-friendly and interactive python implementation which will be pushed to rabit code repo than introducing rabit explicitly to xgboost repo

@tqchen  your opinion?

xydrolase commented:
@CodingCat I was actually wondering myself this morning when submitting this PR, whether this PR fits better in the rabit project or not. 

Let me offer a bit of perspective about why I was motivated to write Rabit tracker in Scala to begin with. The Hadoop cluster we have here have thousands of nodes, some of which happen to have lower version of Python installed. Occasionally, if the driver node happens to be a node with Python  older than 2.6, the tracker would crash. To prevent such randomness in failures due to Python versions, to me, the better option to get rid of the Python implementation. 

It sure adds the complexity of maintenance because the implementations using different languages must be kept in sync. But since I would like to avoid running into Python-related issues, I'd be happy to keep the Scala implementation updated.

xydrolase commented:
Also, with a proper main class, the RabitTracker implemented in Scala should be good on its own, either as a library, or as a standalone tracker service. The only library dependency it has is Akka. Therefore, I don't see any issue moving the implementation to [dmlc/rabit](https://github.com/dmlc/rabit) as well.

@tqchen @CodingCat I'm all open to your suggestions.

CodingCat commented:
@xydrolase thanks for the explanations

If I understand correctly, the most difficulties come from the way in which we treat python-based rabit ...we start it and assume that everything will be OK, and then we start JVM routines which on a healthy rabit 

How about we make python-based rabit more open to the external side relying on it...e.g. we start rabit process from JVM, and after 10 seconds, JVM processes checks the log file to see whether a message like "successful started" has been printed.... or after 10 seconds, we look for a line like "all workers have connected"

to resolve the issue like job crashes due to the old python version....I prefer to leave this logic to the user, e.g. he/she may want to restart manually, or have a loop to keep trying 

Did I miss anything?

xydrolase commented:
@CodingCat Thanks for the reply. The Python-based rabbit tracker does in fact print out information like worker connected, all workers started etc. And on that regard, it is certainly possible to read from the pipe too.

However there still is a considerable risk, at least in our case, that the driver program for Spark, which runs the tracker, would fail RANDOMLY. Spark will retry the driver program. Still, the issue would persist if the driver were allocated to a node with older Python to begin with.
When submitting a distributed training job to the YARN, we launch up to hundreds of workers. When the server is busy, the request for nodes may be queued and it would be annoying if, after enough queueing, the training job ultimately fails due to Python tracker issues. (And the next submission by the user wouldn't guarantee a success because of the inherent randomness.)

Packaging a Python 2.7 or 3.5 within the jar and deploy it at runtime would perhaps resolve this issue but it seems an overkill to me, to make an overblown jar just to run the tracker program.

I understand that the issue pertaining Python issues may be very specific to our use cases. Personally I would prefer xgboost4j to be a pure JVM program so that once I compile `libxgboost4j.so` on the cluster, and compile the jar, it is guaranteed to run. After all, these clusters were mostly for running Hadoop and JVM-based Scala jobs (even PySpark is not supported that well.)

I can test the code thoroughly in terms of performance, functionality etc. And if deemed necessary,  I can move this PR to the rabit project as well, as a separate package.

CodingCat commented:
I am going to discuss with @tqchen about the release plan of XGBoost, and then decide when to merge it...but I prefer to marking it as an experimental component which provides more tools for debugging and graceful handling of network issues 

I would like to keep python-based rabit tracker in the code base so that the user can switch into that once this scala-based one has some problem 

Besides all of these, I would like to see more test cases based on akka-testkit

xydrolase commented:
@CodingCat Thanks with the update. I think it's a good idea to be cautious and mark the Scala-tracker as experimental before it's fully battle-tested. I can modify the code accordingly so that the user can switch the tracker.

Glad to hear that you're discussing with @tqchen on this issue. After more a thorough thought-process, I would say that it probably is better to move the tracker to either `rabit` or `dmlc_core`. Right now, putting a lot of test cases, test scripts etc into XGBoost repository indeed seems a bit weird. 

In the time being, I'll continue working on test cases and akka-testkit. I've tested the rabit client's Allreduce and checkpoint tests. I can also incorporate your suggestions about parsing the output of the Python tracker in the Java code. 

Thanks.

tqchen commented:
Thanks for the PR!  I think as long as we have an experimental option to switch it on it, it is fine to merge it in, but I i will leave the decision to @CodingCat 

Once the new tracker has been proven to work well, we can consider to bring it to dmlc/rabit

xydrolase commented:
@tqchen Thanks for weighing in. I actually like the two-step plan that you're proposing. I'll add the new tracker as an experimental feature to `xgboost4j-spark` and `xgboost4j-flink`.

CodingCat commented:
Hi, @xydrolase I have merged https://github.com/dmlc/xgboost/pull/1608 

before you work on `add options to switch between Python-based tracker and Scala-based tracker in xgboost4j-spark and xgboost4j-flink.` you might want to sync with the master branch

xydrolase commented:
@CodingCat Thanks for the reminder. I'll remind myself about that.

xydrolase commented:
@CodingCat 
All the test cases are complete, and the Scala tracker seems to work properly.
I've also modified the interface of `xgboost4j-spark` to allow user to choose between Java and Scala implementations. Let me know if you have any concerns when code reviewing the interface changes. Thank you in advance.

P.S. There seems to be some issues with Travis CI on R tests at this moment, which caused the checks to fail.

CodingCat commented:
thanks, I am reviewing it, but it may take several days to finish

xydrolase commented:
Thanks. No hurry.

xydrolase commented:
Thanks for the review. For some reason I just got all your comments all in one gigantic batch. 
I'll adopt your comments and suggestions in the coming days.

CodingCat commented:
that's their new feature....write all comments and send all in once...

xydrolase commented on jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/Rabit.java:
```diff
@@ -92,4 +131,73 @@ public static int getWorldSize() throws XGBoostError {
     return out[0];
   }
+
+  public static float[] allReduce(float[] elements, OpType op) {
```
That was purely for the unit test actually. I wanted to test the Rabit tracker minimally possible. Originally I wanted to write the test cases in `xgboost4j`, but somehow `allReduce` can't be called in separate threads so I ended up moving the unit tests to `xgboost4j-spark`, utilizing Spark's executor/container to handle multiple processes for me.

CodingCat replied:
I think you can left only one of them and comment something like purely for test?

xydrolase commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala:
```diff
@@ -0,0 +1,158 @@
+    handler ? RabitTrackerHandler.StartTracker(
+      new InetSocketAddress(InetAddress.getLocalHost,
+        // randomly allocate an ephemeral port if port is not specified
+        port.getOrElse(new Random().nextInt(61000 - 32768) + 32768)
```
The `RabitTrackerHandler` class will handle the sequential probing of available port numbers, if the initial random one is already occupied. See https://github.com/xydrolase/xgboost/blob/c6addeb7a19a0bbc4c62f22cde0294b062303f6a/jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/handler/RabitTrackerHandler.scala#L105

CodingCat replied:
then, would you consolidate the logic about port in a single place?

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  def start(): Boolean = start(30 seconds)
+  def start(timeout: Long, unit: TimeUnit): Boolean = start(Duration(timeout, unit))
```
I think it is just fine to change that....

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala:
```diff
@@ -0,0 +1,158 @@
+    Try(Await.result(futureCompleted, atMost)) match {
+      case Success(n) if n == numWorkers =>
+        system.shutdown()
+        0
```
to obey with the best practice in Java, you may want to replace 0, 1, 2 with enums

xydrolase replied:
Sure.

xydrolase commented on jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/IRabitTracker.java:
```diff
@@ -0,0 +1,20 @@
+  // for Scala-implementation compatibility
+  boolean start(long timeout, TimeUnit unit);
+  int waitFor();
+  int waitFor(long timeout, TimeUnit unit);
```
There are in fact three timeouts.

`start(timeout, unit)` times out for binding to a valid IP address/port (although this rarely happens.)
`waitFor(timeout, unit)` times out waiting for the training tasks.
And the timeout value in the Scala's RabitTracker's constructor times out waiting for workers to connect (typically indicative of network issues, e.g. firewall).

Maybe I should drop the first timeout value, and move the constructor timeout to the `start()` function.

xydrolase replied:
I removed the first one from the `start()` method, which is now in the constructor with a default value. 

The third one (connection timeout) is actually highly dependent on the training data size. 
Because the tracker is started before calling `buildDistributedBoosters`, and the `mapPartitions` in that function is dependent on data loading, which is lazily executed. 
Larger datasets take over 10 minutes to load, and therefore the Rabit workers in `mapPartitions` will start to connect 10 minutes after the tracker has been started. 

As for the second one, I think syntactically it looks more intuitive to have `waitFor()` with a duration argument, instead of specifying it in the constructor. 

To simplify the interface under these considerations, I only kept `start(workerConnectionTimeoutMillis)` and `waitFor(taskExecutionTimeoutMillis)` in `IRabitTracker` interface, and if the timeout is zero, it waits indefinitely.

xydrolase replied:
Really? My personal experience with loading RDDs and then training is that `Rabit.init(rabitEnv.asJava)` runs after loading the RDDs from HDFS, because how Spark constructs the stages as DAGs. I could be wrong though. My experience was based on the log files: the Rabit tracker didn't show any worker connections until all the RDD loading stages are finished. Maybe it was because I added many intermediate stages between data loading and training: I used customized LZO-compressed libsvm file loader to load data, and then use a stratified sampler to transform the RDD, before calling `train()` on XGBoost. 

Okay, I can remove that constructor argument, and make it a hard-coded value of 5 minutes or something.

CodingCat replied:
I see what you mean, if you didn't apply any `action` before XGBoost.train() it may trigger the time-consuming I/O operations.

When using with DataFrame, it might be a rare case. IIRC,  many feature transformers will trigger Spark Jobs 

then this configuration becomes a bit tricky, it highly depends on I/O speed of the users' machines. I will think about it

CodingCat replied:
My suggestion is 

* move `val tracker = startTracker(nWorkers)` to `buildDistributedBoosters`
* check `trainingData`'s STORAGE_LEVEL, if it is not cached, cache `partitionedTrainingSet` in `buildDistributedBoosters`; if it is cached, leave partitionedTrainingSet to save memory
* enforce a Spark job over `partitionedTrainingSet` before moved `val tracker = startTracker(nWorkers)`
* unpersist `partitionedTrainingSet` after a successful training

xydrolase commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala:
```diff
@@ -0,0 +1,158 @@
+  val system = ActorSystem.create("RabitTracker")
+  val handler = system.actorOf(RabitTrackerHandler.props(
+    numWorkers, maxPortTrials, workerConnectionTimeout), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(1 second)
```
Yes, in an implicit fashion for the ask function (`?`).
Although it is a good idea to unify all of the these wait/ask timeouts.

CodingCat replied:
yeah, you can just use askTimeout.duration in all other places

CodingCat commented:
just ping me when it is ready for the further review

xydrolase commented:
@CodingCat Sure. Will finish later today.

xydrolase commented:
@CodingCat If I didn't omit anything, I should have addressed all the comments in your code review. 
You can start reviewing the new changes. 

Thanks.

CodingCat commented:
thanks, I will start it tmr

xydrolase commented:
Thanks. In the meantime, I will think about the suggestions you made about the handling of trainingData. I didn't see them earlier because they were folded.

CodingCat commented:
After giving the issue of training timeout a second thought, I think it might not be necessary to introduce such a variable...essentially, the purpose of this configuration entry is to avoid infinite training caused by lost workers...for XGBoost which has a generally short training time, we can just kill the job with a lost worker (and optionally restart the job)

By removing this variable, we do not need to enforce a Spark job which might seems a bit weird to the user

xydrolase commented:
But wouldn't enforcing a timeout benefit the use case of recurrent/automatic model training?

Say, if I want to study the performance decay of the same gradient boost tree model over time, I train a new model using fresh data, say, every day or week, and compare the validation error to an existing model trained with older data. 
The pipeline for such training tasks are automatic and are triggered by YARN or Oozie RESTful APIs, and if the training tasks timeout, proper alerts and retry logic can be applied. 
Otherwise, I'll have to manually monitor the tasks for potential failures by looking at the console log, and figure out the cause of prolonged training sessions. 

If the users does not care about timeout logic, they can simply ignore the "tracker_conf" parameter. And the default parameter allows the Spark job to run forever.

CodingCat commented:
the question is `what's the reason for a timeout`....

* if it is due to the high time complexity of training tasks, they have multiple ways to monitor those long-running Spark jobs (I once look at a search engine's cluster log, those long-running jobs will be killed by cluster manager (something like Borg) automatically). In this case, we do not need to handle it in a machine learning library and just leave it to the cluster management frameworks

* if it is caused by a lost worker, we just kill or restart the training job before the timeout happens to keep our implementation as simple as possible

xydrolase commented:
Personally, I think every single resource-consuming job should have a timeout just so that it can gracefully exit, without relying on external forces to stop the job.

I can certainly see the alternative solution to use external monitor and management tool to kill a job if it runs amok, or too long. Although it is still up to what tools that are available in the environment, and not all of our clusters have provided APIs for managing applications. 

As for the second case. Spark will handle lost executors by replacing a crashed one with a new container. Although apparently the training task will hang because there's no progress recovery in Rabit. Then wouldn't I still have to monitor the events of executor loss either internally or externally, and restart/kill the job? And it appears to be more complex than just setting a timeout value. 

It does seem a bit unusual to have a timeout in a long-running machine learning task, but it is simple, and requires virtually no external setups for monitoring or management.

CodingCat commented:
> Personally, I think every single resource-consuming job should have a timeout just so that it can gracefully exit, without relying on external forces to stop the job.

> I can certainly see the alternative solution to use external monitor and management tool to kill a job if it runs amok, or too long. Although it is still up to what tools that are available in the environment, and not all of our clusters have provided APIs for managing applications.

Let's take Spark as an example, if I didn't miss any important functionality improvement (maybe I did), Spark does not have such a mechanism to control the maximum job running time...Spark is much more general than XGBoost on that it is a data processing framework instead of an implementation of a particular algorithm...regarding this design philosophy, we shall not over-design XGBoost...

> As for the second case. Spark will handle lost executors by replacing a crashed one with a new container. Although apparently the training task will hang because there's no progress recovery in Rabit. Then wouldn't I still have to monitor the events of executor loss either internally or externally, and restart/kill the job? And it appears to be more complex than just setting a timeout value.

> It does seem a bit unusual to have a timeout in a long-running machine learning task, but it is simple, and requires virtually no external setups for monitoring or management.

The reason I propose to discard timeout setup is that, it either highly depends on the users' I/O power (requires the user to understand this implementation internals), or it requires us to enforce a Spark job (as well as extra RDD cache/unpersist operations) which makes the execution flow of XGBoost weird, and still requires the user to understand the purpose of that extra Spark job....

xydrolase commented:
It is true that Spark or even MLLib do not expose a timeout logic to the user, but barring bugs, Spark handles its internal errors instead of hanging. And it provides UI and tools for monitoring the progress. It is also true that XGBoost itself is unlikely to hang. Hanging is likely caused by abnormal events, like network abnormality, loss of executor, node failure, resource scarcity, resource competition with other running tasks on the cluster etc. Therefore semantically, a timeout does not quite align with the sole purpose of machine learning. 

But it could also be argued that `xgboost4j-spark` is itself a complex framework that builds on top of the backbone of Spark, and many components of `xgboost4j-spark` cannot be directly controlled or monitor by Spark. (For instance, the port numbers of Rabit workers cannot be configured by Spark, and thus the cluster administrator may not add firewall exceptions to allow communications between workers.)

And internally, Spark does have its own heartbeat logic to keep track of the health of workers, and if heartbeat times out too often, a worker/executor is marked as failed, triggering downstream remedies or exceptions. Now that `xgboost4j-spark` has its own communication channels, which are independent of the Spark communication channels, the purpose of the timeouts, in my opinion, is not to enforce a hard threshold on "how long the training task should run", but merely to provide a failsafe if the entire framework weaved by Rabit and Spark runs into issues. And as we add more and more worker nodes to the distributed training effort, the likelihood of having a configuration/network/random issue increases. 

I think it made a lot of sense, per your suggestions, to use `tracker_conf` as a parameter in `paramMap`, instead of using it as an argument in `train` or `trainWithRDD`, because the signatures of the training methods remains unchanged. So again, if the user does not care about timeout, the default behavior resumes to be let XGBoost run forever.

If we were to discard the timeout logic, then I will have to run the training task in a new thread, and implement timeout logic in the driver code. And because the driver is agnostic to what happens internally in the tracker, it cannot timeout when clearly the some workers cannot connect to the tracker, but has to wait the entire timeout duration.

CodingCat commented:
> It is true that Spark or even MLLib do not expose a timeout logic to the user, but barring bugs, Spark handles its internal errors instead of hanging. 

Hanging Spark job is not something uncommon, slow disk write, etc., in usual, users use ctrl+c or kill -9 to kill it, adjust partition number or something else, re-run the job....

> But it could also be argued that xgboost4j-spark is itself a complex framework that builds on top of the backbone of Spark, and many components of xgboost4j-spark cannot be directly controlled or monitor by Spark.

we don't need to rely on Spark to control anything...if it hangs, leave it to the user to resolve the issue just like what Spark does...Just an educated guess, when a user cares a lot about timeout to build an automatic pipeline, they have high probability to equip something to monitor jobs

> the purpose of the timeouts, in my opinion, is not to enforce a hard threshold on "how long the training task should run", but merely to provide a failsafe if the entire framework weaved by Rabit and Spark runs into issues

The issue is that, once you provide such a configurable variable, it becomes `how long the training task should run` already. What's the worse, it implicitly encourages the users to set it to something. Users will use it, then overuse it, and then have a lot of false positive cases to be reported here, e.g. timeout due to I/O workload peaks in data processing phase (suppose that you did not enforce a Spark job before training)

If you enforce a Spark job before training, it makes the whole flow too hacky, but still didn't resolve the issue, e.g. 

1. tracker started in driver side, 

2. all CPU slots in the cluster are occupied, training jobs are struggling for `workerNum` CPUs

3.  unfortunately, timeout before it is able to get enough CPUs and RAMs...

xydrolase commented:
But if a pipeline has multiple stages like training, validation and diagnostics, and say if the diagnostics run considerably longer than the training stage (For instance, testing with adding and dropping features, or a hyper-parameter tuning stage which tune a large set of model parameters.) then the monitoring script has to not only monitor how long the entire pipeline runs, but also how long each pipeline stage runs. And if I have 100 parameters to tune, and the 91st training task hangs, would it be a good option to kill the entire job after finishing 90 parameters, and start over? 
If I allow the 91st training task to timeout, I can simply add a retry logic, or maybe just ignore it, since I have other 99 results anyway. 

Admittedly, adding a new parameter may lead to a small fraction of users to try it and misuse it. I can write detailed documentation/code examples on this, if that's your worry.

xydrolase commented:
Or I can change the timeout so that it is not affected by I/O etc.
Or I can consider removing the timeout in `start()`, but I do want to keep it in `waitFor()` at least, so that the entire training task at least has a timeout, so that sequential training jobs would not hang all together because an individual task hangs (or takes too long. For instance, if I know that I don't want the GBT to be too complex, and I want the training to finish in 30 minutes, then in a grid-search and randomized-search, if the tree has larger `max_depth` and runs too long, I can simply let the training to time out and discard the results from it, keeping the entire parameter tuning stage within a certain time limit.)

CodingCat commented:
> But if a pipeline has multiple stages like training, validation and diagnostics, and say if the diagnostics run considerably longer than the training stage (For instance, testing with adding and dropping features, or a hyper-parameter tuning stage which tune a large set of model parameters.) 
then the monitoring script has to not only monitor how long the entire pipeline runs, but also how long each pipeline stage runs.

I think it is not our problem to resolve, and I didn't see how a timeout in RabitTracker can do better... the user need to set three different values for timeout? Did I miss anything?

> And if I have 100 parameters to tune, and the 91st training task hangs, would it be a good option to kill the entire job after finishing 90 parameters, and start over? 
If I allow the 91st training task to timeout, I can simply add a retry logic, or maybe just ignore it, since I have other 99 results anyway.

Any typo here? You have 100 tasks in a job or you have 100 jobs? I didn't get that how to train 100 models in a single job?

CodingCat commented:
> Or I can change the timeout so that it is not affected by I/O etc.
> Or I can consider removing the timeout in start(), but I do want to keep it in waitFor() at least, so that the entire training task at least has a timeout, so that sequential training jobs would not hang all together because an individual task hangs (or takes too long. For instance, if I know that I don't want the GBT to be too complex, and I want the training to finish in 30 minutes, then in a grid-search and randomized-search, if the tree has larger max_depth and runs too long, I can simply let the training to time out and discard the results from it, keeping the entire parameter tuning stage within a certain time limit.)

I'm not removing timeout from all places...just want to eliminate those over-design...for training timeout, until I see users explicitly say XGBoost shall handle it and convince me with some examples, e.g. TensorFlow will kill a long-running job or Spark will kill a long-running job in themselves instead of relying on job management frameworks, I will take it as an over-design

CodingCat commented:
Consider this in two cases, 

* short training, let's just fail a job in case of a rabit worker (I think Akka IO can monitor TCP connection lost)

* long training.....it becomes tricky for the user to set a good value, who knows what will happen in, e.g. 10 hours, 2 days, some workload peaks will prolong the training job,

xydrolase commented:
> I think it is not our problem to resolve, and I didn't see how a timeout in RabitTracker can do better... the user need to set three different values for timeout? Did I miss anything?

Nope, the whole point is that if a single stage within a pipeline takes too long, I don't have to kill the entire pipeline and start all over. 

> Any typo here? You have 100 tasks in a job or you have 100 jobs? I didn't get that how to train 100 models in a single job?

Let's say I want to tune XGBoost hyperparameters (`max_depth`, `min_child_weight`, `alpha`, `lambda`) on a 4-dimensional-grid. Apparently these parameters will have a significant impact on the overall time of model training, because they severely alter the model complexity. And say that I want to test 100 combinations of hyperparameters, I can run a job that test these 100 hyperparameters sequentially (or in distributed fashion.), but only allow each model training (with different hyperparameter) to run a given amount of time. So that if one model takes too long or hangs, it simply times out. 

Prior to your major overhaul of the xgboost4j-spark interface, I implemented some customized parameter tuning code that do not use Spark's ML Tuning classes and Params, which iterates through, say 100 parameter sets in a single job, and eventually report the optimal parameters with highest score. 
If one out of 100 hangs or times out, I can ignore it and continue the tuning process. Without the timeout logic, all subsequent training tasks will have to wait.

CodingCat commented:
> I think it is not our problem to resolve, and I didn't see how a timeout in RabitTracker can do better... the user need to set three different values for timeout? Did I miss anything?
>> Nope, the whole point is that if a single stage within a pipeline takes too long, I don't have to kill the entire pipeline and start all over.

each of these stages is a single Spark job, when it takes too long, there must be something wrong happening, fail this stage...I didn't see why I need to fail a complete pipeline without a timeout 

> Any typo here? You have 100 tasks in a job or you have 100 jobs? I didn't get that how to train 100 models in a single job?
>> Let's say I want to tune XGBoost hyperparameters (max_depth, min_child_weight, alpha, lambda) on a 4-dimensional-grid. Apparently these parameters will have a significant impact on the overall time of model training, because they severely alter the model complexity. And say that I want to test 100 combinations of hyperparameters, I can run a job that test these 100 hyperparameters sequentially (or in distributed fashion.), but only allow each model training (with different hyperparameter) to run a given amount of time. So that if one model takes too long or hangs, it simply times out.

but you have to run 100 jobs to test each combination. Similar to the above, when a long-running job happens, there must be something wrong. In this case, fail it without the timeout value, why we bothering to set a timeout value and meanwhile introduce the risk of an improper configuration?

xydrolase commented:
> short training, let's just fail a job in case of a rabit worker (I think Akka IO can monitor TCP connection lost)

It can, but typically if the user has a network issue, as reported in #1748 , the worker cannot connect to the tracker, and tracker simply hangs while waiting for all workers to connect. 
After sending "start" command to the tracker, the worker will close connection, and only talk with other workers during `Allreduce`, so there's no open TCP connections for the tracker to monitor. The tracker only sits there waiting for workers to send "print" command, reporting training progress and scores per tree.

> long training.....it becomes tricky for the user to set a good value, who knows what will happen in, e.g. 10 hours, 2 days, some workload peaks will prolong the training job,

It is a bit tricky. But if a user runs training tasks routinely and repetitively, he or she should have at least some sense of a ballpark running time. For instance, if my job takes 6 hours on average, I would want to kill it after 24 hours. Can I kill it manually after 24 hours, absolutely. But if the training job is triggered automatically by, say a Jenkins job, I want the Jenkins job to handle such thing (timeout-related failures.)

Will someone misuse or overuse the timeout and get unexpected results? Probably. But with proper documentation I think the chance is slim.

CodingCat commented:
> short training, let's just fail a job in case of a rabit worker (I think Akka IO can monitor TCP connection lost)

>> It can, but typically if the user has a network issue, as reported in #1748 , the worker cannot connect to the tracker, and tracker simply hangs while waiting for all workers to connect. 
After sending "start" command to the tracker, the worker will close connection, and only talk with other workers during `Allreduce`, so there's no open TCP connections for the tracker to monitor. The tracker only sits there waiting for workers to send "print" command, reporting training progress and scores per tree.

In this case, you are handling with a worker connection timeout instead of training timeout, right?

> long training.....it becomes tricky for the user to set a good value, who knows what will happen in, e.g. 10 hours, 2 days, some workload peaks will prolong the training job,

It is a bit tricky. But if a user runs training tasks routinely and repetitively, he or she should have at least some sense of a ballpark running time. For instance, if my job takes 6 hours on average, I would want to kill it after 24 hours. Can I kill it manually after 24 hours, absolutely. But if the training job is triggered automatically by, say a Jenkins job, I want the Jenkins job to handle such thing (timeout-related failures.)

Will someone misuse or overuse the timeout and get unexpected results? Probably. But with proper documentation I think the chance is slim.

xydrolase commented:
> but you have to run 100 jobs to test each combination. Similar to the above, when a long-running job happens, there must be something wrong. In this case, fail it without the timeout value, why we bothering to set a timeout value and meanwhile introduce the risk of an improper configuration?

I typically run it as 100 stages, instead of 100 separate jobs. So that I don't have to load the data over webHDFS 100 times. 

It is true that some of the single point of failure issue can be avoided if I decouple the stages into separate jobs, but obviously it comes with the additional overhead of I/O due to reloading datasets.

CodingCat commented:
how you run it as stages....since all booster training is triggered by `foreachPartition`

xydrolase commented:
I'm in the process of converting some of the code to the design pattern of using Spark's ML Pipeline.
Since I haven't tested it yet, I'm curious, would each pipeline stage run as a separate job?

CodingCat commented:
each pipeline stage, e.g. feature transformation, will run as a stage or a chain of transformations will be taken as a whole stage

However, when you want to generate a model, e.g. for different parameter combination, you have to (or say Spark ML will ) run a Spark job for each of them...

xydrolase commented:
For instance, if I have a code snippet like this,

```
    smpParams.map { param =>
      val trainingPartitions = Math.max(1, Math.ceil(numPartitions * param.fraction).toInt)
      val subsampledRDD = trainSet.sample(withReplacement, param.fraction, Random.nextLong())
        .coalesce(trainingPartitions, shuffle = true)

      val xgbModel = XGBoost.train(subsampledRDD, configMap, round, trainingPartitions)
      val predictedProba = xgbModel.predict(testFeatures)
      param -> (xgbModel, ModelDiagnostics.areaUnderROC(predictedProba, testLabels))
    }.toMap
```

The closure for the `map` function will run a chain of transformations, but all inside one job. 
Let's say if one of the `XGBoost.train` hangs or takes way too long during the `map` process, can I do anything to kill that training process without having to start over the entire `map` function ?

CodingCat commented:
With the above code, you cannot do it...

how about using Future.sequence to start each of them in a separate Future and you can capture the failed future?

or you can start one by one and skip the failed job to continue the next one in a more imperative-styled code

xydrolase commented:
> how about using Future.sequence to start each of them in a separate Future and you can capture the failed future?

Yeah, but in order for me to not wait _forever_, I guess I will have to wait the Future with a timeout.
I guess that could work, moving the timeout logic out of the XGBoost.

> or you can start one by one and skip the failed job to continue the next one in a more imperative-styled code

Yes, but without timeout, it will not _fail_, but _hangs_. So still I need some sort of timeout mechanism to make sure that the job doesn't stuck during iteration. 

I guess `Future` and `Await` could work.

CodingCat commented:
still back to the question, `what causes hangs`?

from issue list, I see the only case is that

* user start a tracker

* user starts a training job but the worker did not connect to tracker

* tracker hangs there overnight while workers say they finished training 

in this case, I think you have handled it well with a worker connection timeout 

in other cases, like a crashed process, etc, job will fail

xydrolase commented:
> what causes hangs?
If an executor fails, Spark will spawn a new executor as a replacement. You will actually see, in the logfile, that this new executor attempts to connect to the tracker. But then the Rabit worker in that newly spawned executor hangs because it cannot connect to its parent worker, and the neighboring workers in the tree/ring map because none of those workers expect TCP connections from this new worker (they are not aware of the event of peer worker loss, and the tracker doesn't know the executor died too because worker does not maintain an open connection with the tracker once the training starts.)

Allreduce will hang from that point, and the entire training process hangs. (I will need to read Rabit's C++ code more carefully to see how it handles peer worker loss. I expected it to be robust, but the training process did hang.)

I've only seen this once. So the question is whether we handle this hanging behavior externally? In Rabit tracker? Or in Rabit library?

CodingCat commented:
> this new executor attempts to connect to the tracker.

will tracker know it?

xydrolase commented:
> will tracker know it?

It knows. Because it literally receives a "start" command from a new worker. 
I see what you meant, let the tracker handle such unexpected connections, and throw exceptions accordingly. 

Yeah, that could be a more elegant way to deal with executor loss actually.

As for the training timeout to prevent jobs from running **too long**, I can use `Future` and `Await`.

xydrolase commented:
The tracker should know that the "start" command is invalid because all workers have already connected previously. I can add that logic to the Scala tracker, and add some test cases to test it.

CodingCat commented:
yeah, I prefer that to make things simple (for something like XGBoost), as long as worker side code of Rabit does not put much energy to handle this situation....

xydrolase commented:
Working on it. Will ping you once it's done.

xydrolase commented:
I'm trying to implement a unit case by purposefully throwing an exception inside `foreachPartition` from one of the executor. And found that if the worker throws an exception inside `foreachPartition()`, then because no `Rabit.shutdown` was never called, the tracker hangs waiting for shutdown signals. 

I guess one solution is to wrap everything inside the `foreachPartition()` between `Rabit.init` and `Rabit.shutdown` inside a `try{}` block, and calls `Rabit.shutdown` regardless to make sure that the tracker never hangs waiting for "shutdown" commands.

The problem with "shutdown" command is that the command itself is stateless, it doesn't really contain a `statusCode` to indicate whether the shutdown was due to successful execution, or due to an exception. Therefore there's no way to properly reflect that in the return code of `waitFor()`.

Need some further thoughts about how to resolve this hanging issue.

CodingCat commented:
when an exception is thrown by the code defined in the closure in `buildDistributedBoosters`, shouldn't the task be restarted, and the tracker will receive an extra `start`?

CodingCat commented:
I didn't try, but if foreachPartition is failed, the thread starting the job shall throw an exception (no?)

you can 

* make RabitTracker implements UncaughtExceptionHandler
* set UncaughtExceptionHandler of the thread to RabitTracker
* terminate waitFor when receiving a UncaughtException in RabitTracker

CodingCat commented:
> Unit tests apparently can't cover use cases in YARN, where the executor spawning and retry are a bit different. Without reproducible errors to test, it's hard to write tracker code to detect executor loss.

an extra 'start' does not exist?

xydrolase commented:
> when an exception is thrown by the code defined in the closure in buildDistributedBoosters, shouldn't the task be restarted, and the tracker will receive an extra start?

Local mode doesn't have restart by default.
And if you enforce restart by doing something like this:
`val sparkConf = new SparkConf().setMaster("local[*]").`
as I said earlier, you got this error thrown by the native code `rabit::Init is already called in this thread`, which appeared to me crashing the program entirely (because the worker does not run in a container.)

xydrolase commented:
I didn't really change `buildDistributedBoosters` because all I did was trying to write a unit test. I'm reusing my own Rabit Allreduce test code, where I have a similar `foreachPartition()` function call.

Of course I can modify the code of `buildDistributedBoosters ` to purposefully test that, which is what I'll try next.

CodingCat commented:
we don't need to care too much about local mode.....crash the entire program is just fine

xydrolase commented:
> I didn't try, but if foreachPartition is failed, the thread starting the job shall throw an exception (no?)

I'll test `buildDistributedBoosters` to see how the errors are handled there.

xydrolase commented:
> we don't need to care too much about local mode.....crash the entire program is just fine

I don't mind it crashing, better than hanging. 😄 
It's just that local mode kind of hinders the testing of executor respawning.

xydrolase commented:
`UncaughtExceptionHandler` shoud do the trick. Will test it properly later tonight.

xydrolase commented:
@CodingCat I've removed the `trainingTimeout` from `TrackerConf`. Although I did't revert the interface changes on IRabitTracker. After all, the end-user does not directly use tracker. 

Keeping the timeout logic inside the tracker makes sure that if the Scala tracker later becomes an independent library, at least it has built-in timeout logic.

I've added handling of exceptions and invalid worker "start" commands in the tracker too. 
You can start reviewing at your earliest convenience.

Thanks!

CodingCat commented on jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRobustnessSuite.scala:
```diff
@@ -0,0 +1,136 @@
+  test("test Scala RabitTracker's workerConnectionTimeout") {
+    val sparkConf = new SparkConf().setMaster("local[*]")
+      .setAppName("XGBoostSuite").set("spark.driver.memory", "512m")
+    implicit val sparkContext = new SparkContext(sparkConf)
```
you can extend `SharedSparkContext`

xydrolase replied:
See the comment for creating the first `sparkContext`. I did initially extend `SharedSparkContext`, and realized that if I reuse the Spark context, it will reuse the same Thread pool locally. Because I am literally simulating worker exception here, it means that `Rabit.shutdown()` will never be called, hence the internal pointer is not released and set to `nullptr` in the C++ code.

As a side effect, you can't call `Rabit.init()` twice within the same thread. I explicitly create new Spark context to force creating a new Thread pool.

CodingCat commented on jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRobustnessSuite.scala:
```diff
@@ -0,0 +1,136 @@
+      // simulate that the first worker cannot connect to tracker due to network issues.
+      if (index != 1) {
+        Rabit.init(trackerEnvs)
+        Thread.sleep(500)
```
maybe a safer value, like 800?

CodingCat commented on jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/IRabitTracker.java:
```diff
@@ -0,0 +1,28 @@
+public interface IRabitTracker extends Thread.UncaughtExceptionHandler {
+  Map<String, String> getWorkerEnvs();
+  boolean start(long workerConnectionTimeout);
+  int waitFor(long taskExecutionTimeout);
```
would you leave a comment here , saying that taskExecutionTimeout does not take any effect in the current XGBoost?

xydrolase replied:
Sure.

xydrolase commented:
Test cases are not stable on Travis CI, causing crashes. Will look into that today.

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/util/RabitTrackerHelpers.scala:
```diff
@@ -0,0 +1,41 @@
+import java.nio.{ByteOrder, ByteBuffer}
+import akka.util.ByteString
+
+object RabitTrackerHelpers {
```
private[rabit]

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/util/LinkMap.scala:
```diff
@@ -0,0 +1,136 @@
+  }
+}
+
+class LinkMap private[scala](numWorkers: Int) {
```
you need to put private[scala] in front of class, otherwise, you're just make cosntructor private instead of the whole class

xydrolase commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala:
```diff
@@ -0,0 +1,193 @@
+  * @param tcpBindTimeout Timeout for the tracker to bind to an InetSocketAddress.
+  */
+class RabitTracker private[scala](numWorkers: Int, port: Option[Int] = None,
+                                  maxPortTrials: Int = 1000, tcpBindTimeout: Duration = 1 minute)
```
Forgot to remove that.

xydrolase commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala:
```diff
@@ -0,0 +1,193 @@
+
+  // a future for XGBoost worker execution
+  val futureCompleted: Future[Int] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestCompletionFuture, askTimeout.duration)
```
The future here is just to mirror a `Promise` object in the Akka actor, because training completion will occur in the future when the Promise is fulfilled. 

It has a `Await.result` here because ask pattern always return a Future. Basically it's a `Future[Future[Int]]` type, which I need to unpack to get the actual `Future`, which `waitFor` will wait for.

xydrolase replied:
The `askTimeout.duration` will mostly likely never matter, because the handling of `RequestCompletionFuture` message inside the tracker actor is a simple echo, returning a Future object.

CodingCat replied:
similar to the above future, I think it is not necessary to be a Future-typed member field of the class....just apply ask pattern in where you block to wait for the completion of the current future

xydrolase commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala:
```diff
@@ -0,0 +1,193 @@
+  val handler = system.actorOf(RabitTrackerHandler.props(numWorkers), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(30 seconds)
+
+  val futureWorkerEnvs: Future[Map[String, String]] = Try(
```
I can't remember the original motive behind this. Maybe there was another layer of timeout logic that I later removed. 

I'll read through this section, and remove it if the future is not necessary.

xydrolase replied:
This is because when the tracker sends `StartTracker` message to the actor, the actor has its internal messaging logic to handle Tcp port binding etc, which is async. 

Only when the port binding succeeds, the tracker can start properly. Therefore the future is used to behold to the "future" event of port binding.

xydrolase replied:
Would it be possible to do so though? 
My understanding of the ask pattern (`?`) is that, like tell (`!`), it does not block. By using `?`, the sender (`RabitTracker`) gets a future back. And because by the time the tracker "asks" the actor, the worker envs may not be ready yet, the actor returns a Future, which is wrapped into another Future automatically by `?`. 

Correct me if I'm wrong here.

CodingCat replied:
I may not read the logic here carefully, I didn't get the reason that why you cannot block on the future returned by `?`

CodingCat replied:
because introduce a member field typed as `Future` seems fishy...essentially such a design idea means that, some of the state of a class is non-determined...and considering that Future is a stateless entity....it seems weird....

also, in this part, a lot of blocking operations are introduced....you may know that...this usually suggests a potentially better design in the context of Akka

xydrolase replied:
If I wait for the future in the constructor, then it blocks in the constructor forever (or until timeout.)
The tracker will not start binding to socket and accepting connections until the user calls `start()`, which sends a `StartTracker` message to the actor. Without `StartTracker` message, the actor does nothing, and the promise is never fulfilled.

CodingCat replied:
no, I didn't say you need to block in constructor....you can block within start()

xydrolase replied:
It was probably because I wanted to make the variables immutable. 
I'll try to fix that by changing some variables to mutable.

CodingCat replied:
it may waste some of your existing efforts, but I doubt whether it is really necessary to use Akka IO in RabitTracker case...

The complexity here comes from the fact that binding to a IP and port is asynchronous operation.....(after you send StartTracker, the handler of StartTracker send a Tcp.Bind expecting  the handler of Tcp.Bound to fill `promisedWorkerEnvs `)

If you consider what's the usual case of Akka IO or similarly, Spray, most of them are handling a large number of Tcp Connections, so asynchronous binding helps a lot

In our case, only one binding is needed....

xydrolase commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala:
```diff
@@ -0,0 +1,193 @@
+    * @return HashMap containing tracker information.
+    */
+  def getWorkerEnvs: java.util.Map[String, String] = {
+    new java.util.HashMap((Await.result(futureWorkerEnvs, 0 nano) ++ Map(
```
The Future must have already succeeded by the time the user calls `getWorkerEnvs`, therefore there's no need to have any timeout at all. 

To echo your comment above, probably this Future is not needed.

CodingCat commented:
I will go to sleep...drunk a bit too more....will get back tmr

xydrolase commented:
@CodingCat  Have a good weekend!
I've removed the future logic from constructor per your suggestions.

The code is ready for further review.

CodingCat commented:
Sorry for the delay, i am moving to US...full schedule these days

Hope I can get back soon

xydrolase commented:
@CodingCat No problem at all. Congrats.

CodingCat commented:
it seems that the test is still flaky?

xydrolase commented:
It appears so. The linux dev machine I used for compilation and tests have only two cores.
The Travis-CI node has 32-cores, and will switch to a 8-core one and see if I can reproduce the errors.

xydrolase commented:
It is due to the fact that once an exception is thrown, the UncaughtExceptionHandler sends signal to the tracker actor to shut it down, thus all future worker connections will fail. 

In a proper cluster mode, worker will fail inside the container of an executor; On local mode however, it causes the entire process to halt, and thus aborted the test. 

I can either play with the the delays to see if it helps, or I can reduce the number of cores in the testing environment (restrict it to 2--4 cores say.)

xydrolase commented:
@CodingCat Tried some tactics to make the test cases more stable by throwing the exception in the last worker. Maybe we should let Travis CI to perform checks multiple times just to make sure that the build is stable "enough"?

CodingCat commented:
I didn't get why limit core number can fill the gap between Local and distributed mode of Spark in this case

I found a fishy line `java: cxa_atexit.c:100: __new_exitfn: Assertion `l != ((void *)0)' failed.`  in the log...it looks like a core dump

so, maybe the real reason is that Process.destroy is not thread safe?

xydrolase commented:
Limiting core number doesn't technically "fill the gap", but reducing the number of workers (threads) on local mode.

I think the Assertion failure you quoted was in librabit: after retrying connecting to tracker a few times without a success, the Rabit client will exit the program, and thus kill all the threads. 

Since in local mode, all workers are merely threads and reside in the same process as the driver/tracker/etc, one worker exiting the program kills the entire Spark application, hence the test didn't even finish. 

I didn't end up reducing the number of cores. I requested the last worker to throw an exception, therefore it is less likely that a worker cannot connect to a tracker that has already shut down.

CodingCat commented:
I still have questions on it, since tracker is a forked process, why it will stop the Spark application in the case of a failure worker?

xydrolase commented:
The crash of the Spark application was not **directly** caused by killing the forked tracker process. 
However, as we kill the tracker prematurely while there are still workers trying to connect to the tracker, the workers will fail to connect.

Once the Rabit worker fails to connect, it will exit the program immediately. And since the Rabit workers were run in the main process, killing the process containing Rabit workers is equivalent to killing the Spark application.

CodingCat commented:
hmmm....I see, it makes sense

you may want to add some comments to explain why the test cases look like that

xydrolase commented:
Sure, I can do that.

CodingCat commented:
others LGTM

xydrolase commented:
@CodingCat Comments added.

CodingCat commented:
sure, will merge after Travis done
